File: Mastering matlab-6/writemat.f
C 
C   writemat.f - Create a binary MAT file.
C 
C   Mastering MATLAB MAT-file Example
C 
C  

C    B.R. Littlefield, University of Maine, Orono, ME 04469
C    7/20/00
C    Mastering MATLAB 6, Prentice Hall, ISBN 0-13-019468-9
C  

      program writemat
C--------------------------------------------------------------
C     Pointers.
C
      integer matOpen, mxCreateFull, mxCreateString
      integer matGetMatrix, mxGetPr
      integer mfile, mdata, mstr
C--------------------------------------------------------------
C     Other variables
C     
      integer status, matClose
      double precision dat(12)
      data dat / 1.0, 5.5, -4.0, 
     &           2.0, 6.6, -3.0, 
     &           3.0, 7.7, -2.0, 
     &           4.0, 8.8, -1.0 /
C
C     Open MAT-file for writing.
C
      mfile = matOpen('mmtest.mat', 'w')
      if (mfile .eq. 0) then
         write(6,*) 'Cannot open ''mmtest.mat'' for writing.'
         stop
      end if
C
C     Create the mxArray to hold the numeric data.
C
      mdata = mxCreateFull(4,3,0)
      call mxSetName(mdata, 'mydata')
C
C     Copy the data to the mxArray. 
C
      call mxCopyReal8ToPtr(dat, mxGetPr(mdata), 12)
C
C     Create the string array and set the variable name.
C
      mstr = mxCreateString('Mastering MATLAB Rocks!')
      call mxSetName(mstr, 'mystr')
C
C     Write the mxArrays to the MAT file.
C
      call matPutMatrix(mfile, mdata)
      call matPutMatrix(mfile, mstr)
C
C     Free the mxArray memory.
C
      call mxFreeMatrix(mdata)
      call mxFreeMatrix(mstr)
C
C     Close the MAT file.
C
      status = matClose(mp)
      if (status .ne. 0) then
         write(6,*) 'Cannot close mmtest.mat.'
         stop
      end if
C
      stop
      end

File: Mastering matlab-6/writemat.c
/*
 *  writemat.c - Create a binary MAT file.
 *
 *  Mastering MATLAB MAT-file Example
 *
 */

/*   B.R. Littlefield, University of Maine, Orono, ME 04469
 *   7/20/00
 *   Mastering MATLAB 6, Prentice Hall, ISBN 0-13-019468-9
 */

#include "mat.h"

int makemat(const char *filename, 
            double *data, int m, int n, 
            char *mmstr) 
{
  MATFile *mfile;
  mxArray *mdata, *mstr;
  
  /* Open the MAT file for writing. */
  mfile = matOpen(filename, "w");
  if (mfile == NULL) {
    printf("Cannot open %s for writing.\n", filename);
    return(EXIT_FAILURE);
  }

  /* Create the mxArray to hold the numeric data.     */
  /* Note that the array dimensions are reversed.     */
  /* C uses row order while MATLAB uses column order. */
  /* The data array will be transposed in MATLAB.     */
  mdata = mxCreateDoubleMatrix(n,m,mxREAL);
  mxSetName(mdata, "mydata");

  /* Copy the data to the mxArray. Note that mxGetData is */
  /* similar to mxGetPr in that it returns a void pointer */
  /* while mxGetPr returns a pointer to a double.         */
  memcpy((void *)(mxGetData(mdata)), (void *)data, 
                  m*n*sizeof(double));
  
  /* Create the string array and set the variable name. */
  mstr = mxCreateString(mmstr);
  mxSetName(mstr, "mystr");

  /* Write the mxArrays to the MAT file. */
  matPutArray(mfile, mdata);
  matPutArray(mfile, mstr);

  /* Free the mxArray memory. */
  mxDestroyArray(mdata);
  mxDestroyArray(mstr);

  /* Close the MAT file. */
  if (matClose(mfile) != 0) {
    printf("Cannot close %s.\n",filename);
    return(EXIT_FAILURE);
  }

    return(EXIT_SUCCESS);
}

int main()
{ 
  int status;
  char *mmstr = "Mastering MATLAB Rocks!";
  double data[3][4] = {{  1.0,  2.0,  3.0,  4.0 },
                       {  5.5,  6.6,  7.7,  8.8 },
                       { -4.0, -3.0, -2.0, -1.0 }};
 
  status = makemat("mmtest.mat", *data, 3, 4, mmstr);
  return(status);
}
